feat(android): be able to add plugins to build.gradle

AbdullahFaqeir opened the pull request:
feat(android): added classpaths to manifest
feat(android): add plugins to manifest
feat(android): custom root.build.gradle for modules

- classpaths will be added to the root project build.gradle to allow plugins to be identified.
- plugins will be added to the module build.gradle file.

AbdullahFaqeir committed:
feat(android): added classpathes to manifest
feat(android): plugins classpathes to manifest
feat(android): custom root.build.gradle for modules

hansemannn commented on android/templates/build/root.build.gradle:
```diff
@@ -10,9 +10,19 @@ buildscript {
 		classpath 'com.android.tools.build:gradle:7.0.4'
 		classpath 'com.google.gms:google-services:4.3.10'
 		classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
+		<% for(var i=0; i<classpathes.length; i++) {%>
```
Minor change for better readability:
```suggestion
		<% for (let i = 0; i < classpathes.length; i++) {%>
```

hansemannn reviewed (changes requested):
Just a minor code style issue, thanks!

AbdullahFaqeir committed:
Update root.build.gradle

Co-authored-by: Hans Knöchel <hansemannn@users.noreply.github.com>

AbdullahFaqeir commented:
> Just a minor code style issue, thanks!

Of course! Thanks 🙏

hansemannn reviewed (approved):
CR looks good from my end. @m1ga Any objections?

AbdullahFaqeir commented:
> CR looks good from my end. @m1ga Any objections?

I guess I need to add a couple of more things to make this PR feature solid.

m1ga commented:
i've not tested it as I don't have any test case or knowledge how to use plugins in a gradle file. But if it works for you and doens't break the usual stuff (module build, app build, hyperloop app, kitchensink app) then go for it 😄 
if it is something "non standard" gradle stuff there should be some documentation how to use it

AbdullahFaqeir committed:
feat(android): added default root.build.gradle file

AbdullahFaqeir commented:
@hansemannn @m1ga This was the last commit needed to this PR, I've tested it from my side and it's all good now and ready to be merged, as for the documentation of how to use this, we need to add the update [here](https://titaniumsdk.com/guide/Titanium_SDK/Titanium_SDK_How-tos/Extending_Titanium_Mobile/), if you can tag me on where this is editable so I can work on the documentation of this new feature.

m1ga commented:
Nice, that will be in this repo: https://github.com/tidev/titanium-docs/tree/main/docs/guide/Titanium_SDK/Titanium_SDK_How-tos/Extending_Titanium_Mobile

AbdullahFaqeir commented:
Cool

m1ga commented:
Hyperloop apps won't build.
```
[ERROR] [GRADLE] 
[ERROR] [GRADLE] FAILURE: Build failed with an exception.
[ERROR] [GRADLE] 
[ERROR] [GRADLE] * Where:
[ERROR] [GRADLE] Build file '/home/miga/tools/hyperloop-examples/build/hyperloop/android/build.gradle' line: 2
[ERROR] [GRADLE] 
[ERROR] [GRADLE] * What went wrong:
[ERROR] [GRADLE] Could not compile build file '/home/miga/tools/hyperloop-examples/build/hyperloop/android/build.gradle'.
[ERROR] [GRADLE] > startup failed:
[ERROR] [GRADLE]   build file '/home/miga/tools/hyperloop-examples/build/hyperloop/android/build.gradle': 2: Unexpected input: '{' @ line 2, column 13.
```

The generated build.gradle file looks like this:
```
	dependencies {
		classpath 'com.android.tools.build:gradle:7.0.4'
		classpath 'com.google.gms:google-services:4.3.10'
		classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
		<% for (let i = 0; i < classpathes.length; i++) {%>
		classpath '<%= classpathes[i] %>'
		<% } %>
	}
```
so it didn't replace the loop.

m1ga commented:
Module also won't build:
```
[ERROR] TypeError: Cannot read properties of undefined (reading 'split')
    at AndroidModuleBuilder.generateRootProjectFiles (/home/miga/.titanium/mobilesdk/linux/12.3.0/android/cli/commands/_buildModule.js:518:42)
    at async AndroidModuleBuilder.run (/home/miga/.titanium/mobilesdk/linux/12.3.0/android/cli/commands/_buildModule.js:315:3)
[ERROR] An error occurred during build after 24ms
[ERROR] Cannot read properties of undefined (reading 'split')
```

m1ga commented on android/cli/commands/_buildModule.js:
```diff
@@ -561,6 +563,7 @@ AndroidModuleBuilder.prototype.generateModuleProject = async function generateMo
 	let buildGradleContent = await fs.readFile(path.join(this.moduleTemplateDir, 'build.gradle'));
 	buildGradleContent = ejs.render(buildGradleContent.toString(), {
 		compileSdkVersion: this.compileSdkVersion,
+		plugins: this.manifest.plugins.split(','),
```
check if `this.manifest.plugins` is there before splitting

AbdullahFaqeir replied:
Thanks!

AbdullahFaqeir committed:
Merge branch 'master' into modules-gradle

AbdullahFaqeir reviewed (commented):
check if this.manifest.plugins is there before splitting

AbdullahFaqeir committed:
Update _buildModule.js

check if this.manifest.plugins is there before splitting

AbdullahFaqeir commented:
@m1ga can you please review this, thanks!

m1ga commented on android/cli/commands/_buildModule.js:
```diff
@@ -511,11 +511,13 @@ AndroidModuleBuilder.prototype.generateRootProjectFiles = async function generat
-		path.join(this.buildDir, 'build.gradle'));
+	let buildGradleContent = await fs.readFile(path.join(templatesDir, 'root.build.gradle'));
+	buildGradleContent = ejs.render(buildGradleContent.toString(), {
+		classpathes: this.manifest.classpathes.split(','),
```
please add the same fix here

m1ga commented:
with the fixes I still get 
```
[ERROR] [GRADLE] FAILURE: Build failed with an exception.
[ERROR] [GRADLE] 
[ERROR] [GRADLE] * Where:
[ERROR] [GRADLE] Build file '/home/miga/dev/ti.mlkit/android/build/build.gradle' line: 14
```
when building a module as it will add an empty `classpath` line to the build/build.gradle:
```
buildscript {
	ext.kotlin_version = '1.8.20'

	repositories {
		google()
		mavenCentral()
	}
	dependencies {
		classpath 'com.android.tools.build:gradle:7.0.4'
		classpath 'com.google.gms:google-services:4.3.10'
		classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
		
		classpath ''
		
	}
}
...
```

m1ga commented:
The hyperloop error also still exists. To test it: clone and build https://github.com/tidev/hyperloop-examples

AbdullahFaqeir commented:
@m1ga I'll try to work on it today and finish it ASAP.

hansemannn commented:
@AbdullahFaqeir Any updates here? We'd love to include it in our next release!

AbdullahFaqeir committed:
Merge branch 'master' into modules-gradle

AbdullahFaqeir commented:
> @AbdullahFaqeir Any updates here? We'd love to include it in our next release!

I'll get back and work on this within two days

AbdullahFaqeir committed:
Merge branch 'master' into modules-gradle

AbdullahFaqeir committed:
Merge branch 'master' into modules-gradle

hansemannn commented on android/cli/commands/_build.js:
```diff
@@ -2175,10 +2175,12 @@ AndroidBuilder.prototype.generateRootProjectFiles = async function generateRootP
+	// Generate root "build.gradle" template script to the root build directory.
+	let buildGradleContent = await fs.readFile(path.join(this.templatesDir, 'root.build.gradle'));
+	buildGradleContent = ejs.render(buildGradleContent.toString(), {
+		classpathes: [],
```
Typo? It's classpaths

hansemannn committed:
Merge branch 'master' into modules-gradle

hansemannn committed:
fix: fix typos

hansemannn committed:
fix: fix some more typos

hansemannn commented on android/templates/module/default/template/android/manifest.ejs:
```diff
@@ -15,4 +15,6 @@ name: <%- moduleName %>
 moduleid: <%- moduleId %>
 guid: <%- guid %>
 platform: <%- platform %>
+classpaths:
```
Can you give an example how this would be used? I feel like this should solely be done in the module's gradle files, not the manifest - which is rather meant for top-level metadata.

AbdullahFaqeir replied:
The easiest example and which actually led me to create this PR is that someone from our channel on slack asked if someone can port datadog for android, and when I check the setup for datadog, I found it like this:

```groovy
buildscript {
    dependencies {
        classpath("com.datadoghq:dd-sdk-android-gradle-plugin:x.x.x")
    }
}
plugins {
    id("com.datadoghq.dd-sdk-android-gradle-plugin")
    //(...)
}
android {
    //(...)
}
dependencies {
    implementation "com.datadoghq:dd-sdk-android-rum:x.x.x" 
    //(...)
}

```

You need to add a classpath, and you need to apply a plugin, hopefully this is sufficient answer.

hansemannn replied:
So how would a manifest file look in the above case?

hansemannn replied:
@AbdullahFaqeir Waiting for an example of the manifest, then we can merge

AbdullahFaqeir replied:
classpaths: com.datadoghq:dd-sdk-android-gradle-plugin:x.x.x
plugins: com.datadoghq.dd-sdk-android-gradle-plugin

and the dependency would be used as usual

AbdullahFaqeir replied:
> @hansemannn updated the manifest with how to use.

m1ga commented:
building modules still fails with
```
[ERROR] TypeError: Cannot read properties of undefined (reading 'split')
    at AndroidModuleBuilder.generateRootProjectFiles (/home/miga/.titanium/mobilesdk/linux/12.6.0/android/cli/commands/_buildModule.js:519:40)
    at async AndroidModuleBuilder.run (/home/miga/.titanium/mobilesdk/linux/12.6.0/android/cli/commands/_buildModule.js:316:3)
[ERROR] An error occurred during build after 11ms
[ERROR] Cannot read properties of undefined (reading 'split')
```

and hyperloop example
```
[ERROR] [GRADLE] 
[ERROR] [GRADLE] FAILURE: Build failed with an exception.
[ERROR] [GRADLE] 
[ERROR] [GRADLE] * Where:
[ERROR] [GRADLE] Build file '/home/miga/tools/hyperloop-examples/build/hyperloop/android/build.gradle' line: 1
[ERROR] [GRADLE] 
[ERROR] [GRADLE] * What went wrong:
[ERROR] [GRADLE] Could not compile build file '/home/miga/tools/hyperloop-examples/build/hyperloop/android/build.gradle'.
[ERROR] [GRADLE] > startup failed:
[ERROR] [GRADLE]   build file '/home/miga/tools/hyperloop-examples/build/hyperloop/android/build.gradle': 1: Unexpected input: '{' @ line 1, column 13.
[ERROR] [GRADLE]      buildscript {
[ERROR] [GRADLE]                  ^
[ERROR] [GRADLE]   
[ERROR] [GRADLE]   1 erro
```

because the gradle looks like this:

```
...
                classpath 'com.google.gms:google-services:4.4.1'
                classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
                <% for (let i = 0; i < classpaths.length; i++) {%>
                classpath '<%= classpaths[i] %>'
                <% } %>
        }
```

hansemannn committed:
fix: guard non-existing classpaths

hansemannn commented:
@m1ga I've added a possible fix for that - a missing null-check on the `classpaths` (especially for old modules who don't have the `classpaths:` attribute set in the manifest file. I hope optional chaining is supported in the CLI 😇

AbdullahFaqeir commented:
> @m1ga I've added a possible fix for that - a missing null-check on the `classpaths` (especially for old modules who don't have the `classpaths:` attribute set in the manifest file. I hope optional chaining is supported in the CLI 😇
> 
> Not sure if that also fixes the hyperloop build, but as the other statement has thrown before generating the correct file, it could fix it as well.
> 
> ```js
> // Crashed here
> buildGradleContent = ejs.render(buildGradleContent.toString(), {
>     classpaths: this.manifest.classpaths?.split(',') ?? [],
> });
> 
> // Never reached this code because of the crash, resulting in a template-based gradle file
> await fs.writeFile(path.join(this.buildDir, 'build.gradle'), buildGradleContent);
> ```

I'll go over this, and the review @m1ga asked for.

AbdullahFaqeir committed:
fix(android): fix generating build.gradle files

AbdullahFaqeir commented:
Fixed and tested on a sample [Datadog](https://docs.datadoghq.com/real_user_monitoring/mobile_and_tv_monitoring/setup/android/?tab=java) module.

AbdullahFaqeir committed:
Merge branch 'master' into modules-gradle

AbdullahFaqeir commented:
@m1ga @hansemannn kind reminder

m1ga commented:
A test module built fine but the hyperloop issue is still there. Try building https://github.com/tidev/hyperloop-examples and check if you have the same error ([link](https://github.com/tidev/titanium-sdk/pull/14019#issuecomment-2354773771))

AbdullahFaqeir commented:
@m1ga That's because hyperloop module hook file is copying the root.build.gradle file directly instead of rending it as the module build code does.

https://github.com/tidev/hyperloop.next/blob/d727b188b62887414366ddd44f5357d2bbf51ac2/android/hooks/hyperloop.js#L137-L142

this needs another PR to hyperloop to adapt to this PR.

hansemannn reviewed (approved):
I think this one for the SDK is good to go then! The Hyperloop changes should ideally come soon as well though, so we have a unified solution.

AbdullahFaqeir commented:
@hansemannn I've already opened a PR https://github.com/tidev/hyperloop.next/pull/394

AbdullahFaqeir commented:
@hansemannn I guess I just need to update the manifest file to add comments of how to use it.

AbdullahFaqeir committed:
fix(android): added examples in the manifest file on how to use android classpaths and plugins

AbdullahFaqeir committed:
Merge branch 'modules-gradle' of https://github.com/AbdullahFaqeir/titanium-sdk into modules-gradle
